### DISCUSSION TITLE
Automate / Automate.

### DISCUSSION THREAD

**Discussion by bert269** [toverainc/willow#183] [Q&A]
I know, I am probably jumping the gun here, but I am a lazy person. It seems like during this testing phase, we need to redo these commands over and over and honestly - I'm getting tired of typing.

SO, I came up with a script to run outside of the docker environment - no problem, it works fine,
It's simple and it stops if the previous command fails. Ok, granted - I am not the best coder on the planet, my bash sucks and I am sure there are much better ways to do this, but for me it works:

#!/bin/bash
red=`tput setaf 1`
grn=`tput setaf 2`
rst=`tput sgr0`
clear
r_r () {
v0=0
if [ $v1 -eq $v0 ]
   then
        echo "${grn}Success : "$v2"${rst}"
    else
         echo "${red} * * * * Failed * * * * : "$v2" - Please research and correct${red}"
      exit
fi
}
Sleep 3
echo "Remove the 'willow' directory from system"
cd ~
sudo rm -r -f ~/willow 
v1=$?;v2="willow cleanup complete";r_r
git clone https://github.com/toverainc/willow.git && cd willow
v1=$?;v2="Got Willow downloaded";r_r
sudo ./utils.sh build-docker
v1=$?;v2="docker built";r_r
sleep 5
##### INSERT bash to docker script here ######
sudo ./utils.sh docker
v1=$?;v2="Configuration complete";r_r
cd ~/willow
echo "Please connect ESP-BOX [-lite] to USB port"
sleep 10
echo "Press ENTER when done"
read -s -n 1 key
sudo PORT=/dev/ttyACM0 ./utils.sh erase-flash
v1=$?;v2="First round of 'erase-flash' completed";r_r
sleep 5
sudo PORT=/dev/ttyACM0 ./utils.sh erase-flash
v1=$?;v2="Second round of 'erase-flash' completed";r_r
sleep 5
v1=$?;v2="Flashing device";r_r
sudo PORT=/dev/ttyACM0 ./utils.sh flash

Now: How do I pass the three ./utils.sh commands (install, config and build) to the docker environment, from this original bash script, that is outside the container? (See the "##### INSERT bash to docker script here ######" above)

Please don't hate me for this, I'm just trying to save myself some time.

(and truly, if you want complain about the indentation of the if statement above- do not even bother - I have better things to do with my life than trying to figure out why copy/paste drops the tab in the github environment.) ;)

Thanks

**Comment by kristiankielhofner** [comment]
Our current focus is on dynamic configuration and official binary images ready for easy flashing but if we had to repeat all of those steps for development up to this point we'd be really bothered too! You're doing a lot of things you don't need to. A simple process is:

```git pull```

In container:
```
./utils.sh build
```

From wherever you flash:
```./utils.sh flash-app```

With this combination we can usually go from code update to running on device in two minutes or less (depending on build hardware) with three commands total.

**Reply by bert269** [reply]
Thank you. I totally hear you and do appreciate your feedback.
But I still need to type 3 commands, while I can do it all in one script.
As I said, I am not the best coder and just try to make this work to access these docker commands (./utils.sh install/config/build) from the outside bash script.

**Reply by kristiankielhofner** [reply]
Your script is most of the way there (if it works for you). You can remove the unnecessary commands, tweak the flash command, and do what you're describing. Or am I missing something?

**Reply by bert269** [reply]
Sorry. yes. I am not very clear on what I want to accomplish.

I need to pass: **./utils.sh install && ./utils.sh config && ./utils.sh build && exit**
to the docker command (sudo ./utils.sh docker) - so that I do not have to type it inside the container. The exit at the end will bring me back to the original bash script

**Reply by kristiankielhofner** [reply]
Ah, ok.

What you should probably do is something like what we do for running the container.

So take your script and execute commands for docker with something like:

``` docker run --rm -it -v "$PWD":/willow -e TERM --name "$DOCKER_NAME" "$DOCKER_IMAGE" /bin/bash ./utils.sh build```

Then repeat that for each of your steps inside the container. Obviously any host commands just run on the host. Additionally, the config step isn't required unless you want to change anything in your configuration or change our default options for any new configuration items.

Or you could just write a script for your container commands and execute that instead of ```/bin/bash whatever```.

We decided not to implement a "one shot" script like yours because it's simple enough to do something like it right now and we don't have a more efficient way to determine the user intent vs running the commands separately anyway.

**Reply by bert269** [reply]
It does not like me:
~/willow$ docker run --rm -it -v "$PWD":/willow -e TERM --name "$DOCKER_NAME" "$DOCKER_IMAGE" /bin/bash ./utils.sh config
docker: invalid reference format.
See 'docker run --help'.

docker run --help does not show -it as an option - it has -i : which means:
-i, --interactive                    Keep STDIN open even if not attached

Looking at the code of utils.sh, I see exactly the command you copied from ther, with the -it. I do not know why it does not like what I am trying to do.

Was that the intent?

**Reply by kristiankielhofner** [reply]
Those are variables only set in utils.sh and were meant to be placeholders for you.

For you:

```docker run --rm -it -v "$PWD":/willow -e TERM willow:latest ./utils.sh build```

**Comment by bert269** [comment]
Thanks for your time. It still does not like me:

bert@Ubuntu:~$ sudo docker run --rm -it -v "$PWD":/home/bert/willow -e TERM willow:latest ./utils.sh config
Detecting the Python interpreter
Checking "python" ...
Python 3.8.10
"python" has been detected
Adding ESP-IDF tools to PATH...
Using Python interpreter in /opt/esp/python_env/idf4.4_py3.8_env/bin/python
Checking if Python packages are up to date...
Python requirements from /opt/esp/idf/requirements.txt are satisfied.
Added the following directories to PATH:
  /opt/esp/idf/components/esptool_py/esptool
  /opt/esp/idf/components/espcoredump
  /opt/esp/idf/components/partition_table
  /opt/esp/idf/components/app_update
  /opt/esp/tools/xtensa-esp-elf-gdb/11.2_20220823/xtensa-esp-elf-gdb/bin
  /opt/esp/tools/riscv32-esp-elf-gdb/11.2_20220823/riscv32-esp-elf-gdb/bin
  /opt/esp/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin
  /opt/esp/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin
  /opt/esp/tools/xtensa-esp32s3-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s3-elf/bin
  /opt/esp/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
  /opt/esp/tools/esp32ulp-elf/2.35_20220830/esp32ulp-elf/bin
  /opt/esp/tools/cmake/3.23.1/bin
  /opt/esp/tools/openocd-esp32/v0.11.0-esp32-20221026/openocd-esp32/bin
  /opt/esp/python_env/idf4.4_py3.8_env/bin
  /opt/esp/idf/tools
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:

  idf.py build

**/opt/esp/entrypoint.sh: line 6: /willow/utils.sh: No such file or directory**

It behaves like this for all three options: install, config and build.

Do not spend anymore time on this, at least I'm saving some time by running the script before and after the docker environment, I just have to run inside docker the install, config and build (and exit).

Thanks

**Reply by kristiankielhofner** [reply]
You need to run the command exactly as how I provided it:

```docker run --rm -it -v "$PWD":/willow -e TERM willow:latest ./utils.sh build```

With yours you are mounting the current directory at your home directory path inside the container:

docker run --rm -it -v "$PWD":**/home/bert/willow** -e TERM willow:latest ./utils.sh config

Your command will not work.

**Comment by bert269** [comment]
Makes no difference. I run it from my home directory, where the script lives. Still get the same error message:
bert@Ubuntu:~$ **sudo docker run --rm -it -v "$PWD":/willow -e TERM willow:latest ./utils.sh config**
Detecting the Python interpreter
Checking "python" ...
Python 3.8.10
"python" has been detected
Adding ESP-IDF tools to PATH...
Using Python interpreter in /opt/esp/python_env/idf4.4_py3.8_env/bin/python
Checking if Python packages are up to date...
Python requirements from /opt/esp/idf/requirements.txt are satisfied.
Added the following directories to PATH:
  /opt/esp/idf/components/esptool_py/esptool
  /opt/esp/idf/components/espcoredump
  /opt/esp/idf/components/partition_table
  /opt/esp/idf/components/app_update
  /opt/esp/tools/xtensa-esp-elf-gdb/11.2_20220823/xtensa-esp-elf-gdb/bin
  /opt/esp/tools/riscv32-esp-elf-gdb/11.2_20220823/riscv32-esp-elf-gdb/bin
  /opt/esp/tools/xtensa-esp32-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32-elf/bin
  /opt/esp/tools/xtensa-esp32s2-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s2-elf/bin
  /opt/esp/tools/xtensa-esp32s3-elf/esp-2021r2-patch5-8.4.0/xtensa-esp32s3-elf/bin
  /opt/esp/tools/riscv32-esp-elf/esp-2021r2-patch5-8.4.0/riscv32-esp-elf/bin
  /opt/esp/tools/esp32ulp-elf/2.35_20220830/esp32ulp-elf/bin
  /opt/esp/tools/cmake/3.23.1/bin
  /opt/esp/tools/openocd-esp32/v0.11.0-esp32-20221026/openocd-esp32/bin
  /opt/esp/python_env/idf4.4_py3.8_env/bin
  /opt/esp/idf/tools
Done! You can now compile ESP-IDF projects.
Go to the project directory and run:

  idf.py build

**/opt/esp/entrypoint.sh: line 6: /willow/utils.sh: No such file or directory**

It does not matter if I specifies install, config or build. It looks to me that the docker environment is rebuild each time.
Should it not be a "docker exec" command instead of the "docker run" command?

**Reply by kristiankielhofner** [reply]
Run the provided docker command snippet from the willow directory, wherever it is.

Do not use docker exec.

**Comment by bert269** [comment]
ah-ha. The moment!
That was my problem - I was still in my home directory, trying to run that.

it works, I run the script from my home directory, download the latest willow from git, build-docker, pass control over to the docker environment, run the install, config and build there, return to the willow directory, continue the flashing and it works!

Thanks for your help and sorry about the frustration.

**Reply by kristiankielhofner** [reply]
Glad to hear!
